wandb/sdk/data_types/Image.py

```diff
@@ -0,0 +1,589 @@
+
+        if self._masks is not None:
+            for i, k in enumerate(self._masks):
+                id_ = "{}{}".format(id_, i) if id_ is not None else None
+                self._masks[k].bind_to_run(
+                    run, key, step, id_, ignore_copy_err=ignore_copy_err
+                )
+
+    def to_json(self, run_or_artifact: Union["LocalRun", "LocalArtifact"]) -> dict:
+        json_dict = super(Image, self).to_json(run_or_artifact)
+        json_dict["_type"] = Image._log_type
+        json_dict["format"] = self.format
+
+        if self._width is not None:
+            json_dict["width"] = self._width
+        if self._height is not None:
+            json_dict["height"] = self._height
+        if self._grouping:
+            json_dict["grouping"] = self._grouping
+        if self._caption:
+            json_dict["caption"] = self._caption
+
+        if isinstance(run_or_artifact, wandb.wandb_sdk.wandb_artifacts.Artifact):
+            artifact = run_or_artifact
+            if (
+                self._masks is not None or self._boxes is not None
+            ) and self._classes is None:
+                raise ValueError(
+                    "classes must be passed to wandb.Image which have masks or bounding boxes when adding to artifacts"
+                )
+
+            if self._classes is not None:
+                class_id = hashlib.md5(
+                    str(self._classes._class_set).encode("utf-8")
+                ).hexdigest()
+                class_name = os.path.join("media", "classes", class_id + "_cls",)
+                classes_entry = artifact.add(self._classes, class_name)
+                json_dict["classes"] = {
+                    "type": "classes-file",
+                    "path": classes_entry.path,
+                    "digest": classes_entry.digest,
+                }
+
+        elif not isinstance(run_or_artifact, wandb.wandb_sdk.wandb_run.Run):
+            raise ValueError("to_json accepts wandb_run.Run or wandb_artifact.Artifact")
+
+        if self._boxes:
+            json_dict["boxes"] = {
+                k: box.to_json(run_or_artifact) for (k, box) in self._boxes.items()
+            }
+        if self._masks:
+            json_dict["masks"] = {
+                k: mask.to_json(run_or_artifact) for (k, mask) in self._masks.items()
+            }
+        return json_dict
+
+    def guess_mode(self, data: "np.ndarray") -> str:
+        """
+        Guess what type of image the np.array is representing
+        """
+        # TODO: do we want to support dimensions being at the beginning of the array?
+        if data.ndim == 2:
+            return "L"
+        elif data.shape[-1] == 3:
+            return "RGB"
+        elif data.shape[-1] == 4:
+            return "RGBA"
+        else:
+            raise ValueError(
+                "Un-supported shape for image conversion %s" % list(data.shape)
+            )
+
+    @classmethod
+    def to_uint8(cls, data: "np.ndarray") -> "np.ndarray":
```

Comment:
could be static method since it doesn't depend on the class?

Reply:
I agree, but with such a big change (in terms of surface area, I'd like to not modify any logic)